cachito/workers/pkg_managers/gomod.py

```diff
@@ -286,6 +287,67 @@ def _parse_name_and_version(list_deps_line: str) -> Tuple[str, str]:
     return name, version
 
 
+def _vet_local_deps(dependencies: List[dict], module_name: str, module_path: Path, repo_root: Path):
+    """
+    Fail if any dependency is replaced by a local path unless the module is allowlisted.
+
+    Also fail if the module is allowlisted but the path is absolute or outside repository.
+    """
+    allowed_patterns = None
+
+    for dep in dependencies:
+        name = dep["name"]
+        version = dep["version"]
+
+        if version.startswith("."):
+            if allowed_patterns is None:
+                # Load allowlist only once
+                allowlist = get_worker_config().cachito_gomod_file_deps_allowlist
+                allowed_patterns = allowlist.get(module_name, [])
+                log.debug("Allowed local packages for %s: %s", module_name, allowed_patterns)
+
+            log.debug(
+                "Module %s wants to replace %s with a local dependency: %s",
+                module_name,
+                name,
+                version,
+            )
+            _fail_unless_allowlisted(module_name, name, allowed_patterns)
+            _fail_if_outside_repo(version, module_path, repo_root)
```

Comment:
I found this in the purl spec:
```
When percent-decoded, a segment:

    must not contain a '/'
    must not be any of '..' or '.'
    must not be empty
```
We need a stricter check, not only must the path be inside the repo, it must not start with `..` at all